src/focus-strategies/focus-by-distance.ts

```diff
@@ -0,0 +1,170 @@
+  public primaryDistance = Infinity;
+  public secondaryDistance = Infinity;
+
+  constructor(public el: HTMLElement) {
+    this.rect = this.el.getBoundingClientRect();
+  }
+
+  public calcPercentInShadow(refRect: ClientRect, dir: Direction) {
+    if (isHorizontal(dir)) {
+      this.percentInShadow =
+        Math.min(this.rect.bottom, refRect.bottom) - Math.max(this.rect.top, refRect.top);
+    } else {
+      this.percentInShadow =
+        Math.min(this.rect.right, refRect.right) - Math.max(this.rect.left, refRect.left);
+    }
+  }
+
+  public calcPrimaryDistance(refRect: ClientRect, dir: Direction) {
+    switch (dir) {
+      case Direction.LEFT:
+        this.primaryDistance = refRect.left - this.rect.right;
+        break;
+      case Direction.RIGHT:
+        this.primaryDistance = this.rect.left - refRect.right;
+        break;
+      case Direction.UP:
+        this.primaryDistance = refRect.top - this.rect.bottom;
+        break;
+      case Direction.DOWN:
+        this.primaryDistance = this.rect.top - refRect.bottom;
+        break;
+      default:
+        throw new Error(`Invalid direction ${dir}`);
+    }
+  }
+
+  public calcSecondaryDistance(refRect: ClientRect, dir: Direction) {
+    if (isHorizontal(dir)) {
+      const refCenter = refRect.top + refRect.height / 2;
+      const isAbove = this.rect.bottom < refCenter;
+
+      this.secondaryDistance = isAbove ? refCenter - this.rect.bottom : this.rect.top - refCenter;
+    } else {
+      const refCenter = refRect.left + refRect.width / 2;
+      const isLeft = this.rect.right < refCenter;
+
+      this.secondaryDistance = isLeft ? refCenter - this.rect.right : this.rect.left - refCenter;
+    }
+  }
+}
+
+export class ElementFinder {
+  private shortlisted: PotentialElement[];
+
+  constructor(
+    private dir: Direction,
+    private refRect: ClientRect,
+    candidates: HTMLElement[],
+    private prevEl?: HTMLElement,
+  ) {
+    this.shortlisted = candidates.map(candidate => new PotentialElement(candidate));
+  }
+
+  public find() {
+    this.shortlisted = this.getElementsInDirection();
+
+    if (!this.shortlisted.length) {
+      return null;
+    }
+
+    this.shortlisted.forEach(el => el.calcPercentInShadow(this.refRect, this.dir));
+
+    if (!this.shortlisted.some(el => el.percentInShadow > 0)) {
+      // Case: No elements in shadow
+      //                   +------+
+      //                   |      |
+      //                   +------+
+      // +---------+ --------------
+      // |  X ->   |
+      // +---------+---------------
+      //              +------+   +------+
+      //              |   X  |   |      |
+      //              |      |   |      |
+      //              +------+   +------+
+      if (isHorizontal(this.dir)) {
+        return null;
+      }
+
+      this.shortlisted.forEach(el => el.calcPrimaryDistance(this.refRect, this.dir));
+      const shortestPrimaryDist = this.shortlisted
+        .map(el => el.primaryDistance)
```

Comment:
You can avoid the map here for performance.

Reply:
^ While functional stuff is fast and getting faster, it's still slower than good ol' for loops, and I know you care about performance in many of these paths. Also in some of these scenarios you can do the map+reduce in one iteration instead of two.

Reply:
that sounds like premature optimization considering there are typically <10 shortlisted elements.

Reply:
Yep, sure, apply with pragmatism, haven't read the code thoroughly to determine whether these are appropriate in this case 🙂